corehq/form_processor/steps.py

```diff
@@ -0,0 +1,60 @@
+import re
+from abc import ABC, abstractmethod
+
+from corehq.apps.data_vault.models import VaultStore
+
+
+class FormProcessingStep(ABC):
+    @abstractmethod
+    def __call__(self, *args, **kwargs):
+        """
+        Process the form context object. Return an instance of
+        FormProcessingResult to terminate the processing or None
+        To allow processing to continue to the next step
+        """
+        pass
+
+
+class VaultPatternExtractor(FormProcessingStep):
+    identifier = None
+
+    def __init__(self, patterns, xmlns_whitelist=None):
+        self._patterns = patterns
+        self._xmlns_whitelist = xmlns_whitelist
+
+    def __call__(self, context):
+        if self._should_process(context.instance_xml):
+            return self._process(context)
+
+    def _should_process(self, xml_as_text):
+        for xmlns in self._xmlns_whitelist:
+            if xmlns in xml_as_text:
+                return True
+        return False
+
+    def _process(self, context):
+        new_xml, vault_items = self._replace_values(context.instance_xml)
+        context.instance_xml = new_xml
+        context.supplementary_models.append(vault_items)
+
+    def _replace_values(self, xml_as_text):
+        values = self._extract_patterns_from_string(xml_as_text, self._patterns)
+        distinct_values = set(values)
+        vault_items = [VaultStore(value=value, identifier=self.identifier) for value in distinct_values]
```

Comment:
Ideally the model class should never be used outside of the vault package. The spec suggests creating a python api to interact with the vault store instead of using the models directly. Here's an example of creating a api: https://github.com/dimagi/commcare-hq/pull/28050/files#diff-d4a12607bea635431cd35641c987e855

```
from corehq import vault

vault_values = [vault.create_value(value) for value in distinct_values]
```

Reply:
https://github.com/dimagi/commcare-hq/pull/28161/commits/32deb5f56bd68bcd9f432abaf0d0e4dcbbf75890